contracts/MarketOracle.sol

```diff
@@ -56,12 +65,51 @@ contract MedianOracle is Ownable, IOracle {
         minimumProviders = minimumProviders_;
     }
 
+//    function pushReport(uint256 payload) external
+//    {
+//        address providerAddress = msg.sender;
+//        Report[2] memory reports = providerReports[providerAddress];
+//
+//        require(reports[0].timestamp > 0);
+//
+//        uint8 index_recent = reports[0].timestamp >= reports[1].timestamp ? 0 : 1;
+//        uint8 index_past = 1 - index_recent;
+//
+//        uint256 nowTimestamp = now;
+//
+//        // Check the push is not too soon after the last one.
+//        require(reports[index_recent].timestamp.add(reportSecurityDelaySec) <= nowTimestamp);
+//
+//        providerReports[providerAddress][index_past].timestamp = nowTimestamp;
+//        providerReports[providerAddress][index_past].payload = payload;
+//    }
+
     function pushReport(uint256 payload) external
     {
-        address sender = msg.sender;
-        require(providerReports[sender].timestamp > 0);
-        providerReports[sender].timestamp = now;
-        providerReports[sender].payload = payload;
+        address providerAddress = msg.sender;
+        Report[2] storage reports = providerReports[providerAddress];
+        uint256[2] memory timestamps = [reports[0].timestamp, reports[1].timestamp];
+
+        require(timestamps[0] > 0);
+
+        uint8 index_recent = timestamps[0] >= timestamps[1] ? 0 : 1;
+        uint8 index_past = 1 - index_recent;
+
+        uint256 nowTimestamp = now;
+
+        // Check the push is not too soon after the last one.
+        require(timestamps[index_recent].add(reportSecurityDelaySec) <= nowTimestamp);
+
+        providerReports[providerAddress][index_past].timestamp = nowTimestamp;
+        providerReports[providerAddress][index_past].payload = payload;
+    }
+
+    function purgeReports() external
+    {
+        address providerAddress = msg.sender;
+        require (providerReports[providerAddress][0].timestamp > 0);
+        providerReports[providerAddress][0].timestamp=1;
+        providerReports[providerAddress][1].timestamp=1;
```

Comment:
This writes 1 to both timestamps, but only one is checked in other parts of the code.

Might as well be consistent to avoid any future confusion.

Reply:
For Purge invalidating both reports is needed. But can change other parts to only use report[0].timestamp for checking if a provider is whitelisted.